pkg/backup/backupschedule/backup_schedule_manager.go

```diff
@@ -288,26 +357,232 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
+	}
+
+	var (
+		expiredBackups []*v1alpha1.Backup
+		truncateTSO    uint64
+	)
+
+	if logBackup != nil {
+		expiredBackups, truncateTSO, err = caculateExpiredBackupsWithLogBackup(ascBackups, logBackup, reservedTime)
+		if err != nil {
+			klog.Errorf("caculate expired backups with log backup, err: %s", err)
+			return
 		}
+	} else {
+		expiredBackups, err = caculateExpiredBackups(ascBackups, reservedTime)
+		if err != nil {
+			klog.Errorf("caculate expired backups without log backup, err: %s", err)
+			return
+		}
+	}
+
+	for _, backup := range expiredBackups {
 		// delete the expired backup
-		if err := bm.deps.BackupControl.DeleteBackup(backup); err != nil {
+		if err = bm.deps.BackupControl.DeleteBackup(backup); err != nil {
 			klog.Errorf("backup schedule %s/%s gc backup %s failed, err %v", ns, bsName, backup.GetName(), err)
 			return
 		}
-		deleteCount += 1
 		klog.Infof("backup schedule %s/%s gc backup %s success", ns, bsName, backup.GetName())
 	}
 
-	if deleteCount == len(backupsList) && deleteCount > 0 {
+	if truncateTSO > 0 {
+		// truncate the log backup
+		if err = bm.deps.BackupControl.TruncateLogBackup(logBackup, truncateTSO); err != nil {
+			klog.Errorf("backup schedule %s/%s truncate log backup %s failed, truncateTSO %d, err %v", ns, bsName, logBackup.GetName(), truncateTSO, err)
+			return
+		}
+		klog.Infof("backup schedule %s/%s truncate log backup %s success, truncateTSO %d", ns, bsName, logBackup.GetName(), truncateTSO)
+	}
+
+	if len(expiredBackups) == len(backupsList) && len(expiredBackups) > 0 {
 		// All backups have been deleted, so the last backup information in the backupSchedule should be reset
 		bm.resetLastBackup(bs)
 	}
 }
 
+// separateSnapshotBackupsAndLogBackup return snapot backups ordry by create time asc and log backup
+func separateSnapshotBackupsAndLogBackup(backupsList []*v1alpha1.Backup) ([]*v1alpha1.Backup, *v1alpha1.Backup) {
+	var (
+		ascBackupList = make([]*v1alpha1.Backup, 0)
+		logBackup     *v1alpha1.Backup
+	)
+
+	for _, backup := range backupsList {
+		if backup.Spec.Mode == v1alpha1.BackupModeLog {
+			logBackup = backup
+			continue
+		}
+		ascBackupList = append(ascBackupList, backup)
+	}
+
+	for i := 0; i < len(ascBackupList)-1; i++ {
+		for j := i + 1; j < len(ascBackupList); j++ {
+			if ascBackupList[i].CreationTimestamp.Unix() > ascBackupList[j].CreationTimestamp.Unix() {
+				ascBackupList[i], ascBackupList[j] = ascBackupList[j], ascBackupList[i]
+			}
+		}
+	}
+	return ascBackupList, logBackup
+}
+
+func caculateExpiredBackupsWithLogBackup(backupsList []*v1alpha1.Backup, logBackup *v1alpha1.Backup, reservedTime time.Duration) ([]*v1alpha1.Backup, uint64, error) {
+	var (
+		lastestTSO               uint64
+		expiredTSO               uint64
+		expiredBackups           []*v1alpha1.Backup
+		truncateTSO              uint64
+		isTruncateTSOInLogBackup bool
+		err                      error
+	)
+
+	// caculate latest ts
+	lastestTSO, err = calculateLatestTSO(backupsList, logBackup)
```

Comment:
Shouldn't use lastestTSO, we should use NOW, if backup and logbackup stop, we can not clean the expired data

Reply:
ok, after a discuss, i refine caculate logic and add notes.
https://github.com/pingcap/tidb-operator/pull/4904/files#diff-6d32e105cb814eefb99d8eaa18a5c94f249c6c1ca4e7a9b77b33f6ad218dd479R440